src/x509.c

```diff
@@ -4767,68 +4767,112 @@ WOLFSSL_X509_NAME* wolfSSL_X509_get_subject_name(WOLFSSL_X509* cert)
 
 #if defined(OPENSSL_EXTRA) && (!defined(NO_SHA) || !defined(NO_SHA256))
 /******************************************************************************
-* wolfSSL_X509_subject_name_hash - compute the hash digest of the raw subject name
-* This function prefers SHA-1 (if available) for compatibility
+* wolfSSL_X509_subject_name_hash
+* wolfSSL_X509_issuer_name_hash
+* Compute the hash digest of the subject / issuer name.
+* These functions prefers SHA-1 (if available) for compatibility.
 *
 * RETURNS:
-* The beginning of the hash digest. Otherwise, returns zero.
+* The first 4 bytes of SHA hash in little endian order as unsigned long.
+* Otherwise, returns zero.
 * Note:
-* Returns a different hash value from OpenSSL's X509_subject_name_hash() API
-* depending on the subject name.
+* Returns the same hash value as OpenSSL's X509_X_name_hash() API
+* if SHA-1 support is compiled in.
 */
 unsigned long wolfSSL_X509_subject_name_hash(const WOLFSSL_X509* x509)
 {
-    unsigned long ret = 0;
-    int retHash = NOT_COMPILED_IN;
-    WOLFSSL_X509_NAME *subjectName = NULL;
-    byte digest[WC_MAX_DIGEST_SIZE];
+    unsigned long      hash = 0;
+    WOLFSSL_X509_NAME* subjectName = NULL;
+    unsigned char*     canonName = NULL;
+    byte               digest[WC_MAX_DIGEST_SIZE];
+    int                size = 0;
 
     if (x509 == NULL) {
-        return ret;
+        return 0;
     }
 
     subjectName = wolfSSL_X509_get_subject_name((WOLFSSL_X509*)x509);
-    if (subjectName != NULL) {
+
+    if (subjectName == NULL) {
+        WOLFSSL_MSG("wolfSSL_X509_get_subject_name error");
+        return 0;
+    }
+
+    size = wolfSSL_i2d_X509_NAME_canon(subjectName, &canonName);
+
+    if (size <= 0 || canonName == NULL){
+        WOLFSSL_MSG("wolfSSL_i2d_X509_NAME_canon error");
+        return 0;
+    }
+
     #ifndef NO_SHA
-        retHash = wc_ShaHash((const byte*)subjectName->name,
-                             (word32)subjectName->sz, digest);
+    if (wc_ShaHash((const byte*)canonName, (word32)size, digest) != 0) {
+        WOLFSSL_MSG("wc_ShaHash error");
```

Comment:
canonName will leak here and in other error cases below. @SparkiDev and I (and others, probably) prefer to have one return site per function, which helps with this kind of memory leak error. If you have only a single return site, you only have to free memory in one place, too.